subsys/net/lib/sockets/getaddrinfo.c

```diff
@@ -111,43 +124,31 @@ int z_impl_z_zsock_getaddrinfo_internal(const char *host, const char *service,
 	/* Link entries in advance */
 	ai_state.ai_arr[0].ai_next = &ai_state.ai_arr[1];
 
-	/* Execute if AF_UNSPEC or AF_INET4 */
-	if (family != AF_INET6) {
-		ret = dns_get_addr_info(host, DNS_QUERY_TYPE_A, NULL,
-					dns_resolve_cb, &ai_state, 1000);
-		if (ret == 0) {
-			k_sem_take(&ai_state.sem, K_FOREVER);
-			st1 = ai_state.status;
-		} else {
-			errno = -ret;
-			st1 = DNS_EAI_SYSTEM;
+	ret = exec_query(host, family, &ai_state);
+	if (ret == 0) {
+		/* If the DNS query for reason fails so that the
+		 * dns_resolve_cb() would not be called, then we want the
+		 * semaphore to timeout so that we will not hang forever.
+		 * So make the sem timeout longer than the DNS timeout so that
+		 * we do not need to start to cancel any pending DNS queries.
+		 */
+		int ret = k_sem_take(&ai_state.sem,
+				     CONFIG_NET_SOCKETS_DNS_TIMEOUT +
+				     K_MSEC(100));
```

Comment:
This is an issue (though not what is causing my crash).
CONFIG_NET_SOCKETS_DNS_TIMEOUT as it's passed in here is used for each resolver attempt.  We could in theory have several DNS servers defined, so this value is used several times for an overall total of:
CONFIG_NET_SOCKETS_DNS_TIMEOUT * SERVER_COUNT (DNS + MDNS servers)

Reply:
The resolver does not wait all the results from all the servers, it just waits results from at least one server. So there should be no issue here regarding this.

Reply:
You are correct.  I looked closely at the resolve.c code.  It only continues iterating through the DNS_SERVER list *if* there's a problem in dns_write().  As soon as it hits a successful return value it stops and waits for results or a timeout.